src/libraries/System.Text.Json/src/System/Text/Json/Writer/Utf8JsonWriterCache.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+
+namespace System.Text.Json
+{
+    /// <summary>
+    /// Defines a thread-local cache for JsonSerializer to store reusable Utf8JsonWriter/IBufferWriter instances.
+    /// </summary>
+    internal static class Utf8JsonWriterCache
+    {
+        [ThreadStatic]
+        private static ThreadLocalState? s_threadLocalState;
+
+        public static Utf8JsonWriter RentWriterAndBuffer(JsonSerializerOptions options, out PooledByteBufferWriter bufferWriter)
+        {
+            ThreadLocalState state = s_threadLocalState ??= new();
+            Utf8JsonWriter writer;
+
+            if (state.RentedWriters++ == 0)
+            {
+                // First JsonSerializer call in the stack -- initialize & return the cached instances.
+                bufferWriter = state.BufferWriter;
+                writer = state.Writer;
+
+                bufferWriter.InitializeEmptyInstance(options.DefaultBufferSize);
+                writer.Reset(bufferWriter, options.GetWriterOptions());
+            }
+            else
+            {
+                // We're in a recursive JsonSerializer call -- return fresh instances.
+                bufferWriter = new PooledByteBufferWriter(options.DefaultBufferSize);
+                writer = new Utf8JsonWriter(bufferWriter, options.GetWriterOptions());
+            }
+
+            return writer;
+        }
+
+        public static Utf8JsonWriter RentWriter(JsonSerializerOptions options, PooledByteBufferWriter bufferWriter)
+        {
+            ThreadLocalState state = s_threadLocalState ??= new();
+            Utf8JsonWriter writer;
+
+            if (state.RentedWriters++ == 0)
+            {
+                // First JsonSerializer call in the stack -- initialize & return the cached instance.
+                writer = state.Writer;
+                writer.Reset(bufferWriter, options.GetWriterOptions());
+            }
+            else
+            {
+                // We're in a recursive JsonSerializer call -- return a fresh instance.
+                writer = new Utf8JsonWriter(bufferWriter, options.GetWriterOptions());
+            }
+
+            return writer;
+        }
+
+        public static void ReturnWriterAndBuffer(Utf8JsonWriter writer, PooledByteBufferWriter bufferWriter)
+        {
+            Debug.Assert(s_threadLocalState != null);
+            ThreadLocalState state = s_threadLocalState;
+
+            writer.ResetAllStateForCacheReuse();
+            bufferWriter.ClearAndReturnBuffers();
```

Comment:
I see... so we're pooling the buffer writer instance itself but it in turn is not holding on to any buffers while pooled, is that correct?  So we're effectively just pooling two relatively small objects?

Reply:
That's right 👍